src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * This file is part of Dependency-Track.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) OWASP Foundation. All Rights Reserved.
+ */
+package org.dependencytrack.model;
+
+/**
+ * Defines various types of logics to be applied to collection projects.
+ * Collection projects don't contain own components, instead collect their metrics and
+ * data from their children. The logic to apply when calculating this data is defined
+ * by this type.
+ *
+ * @author Ralf King
+ * @since 4.11.0
+ */
+public enum ProjectCollectionLogic {
+    /**
+     * Project is not a collection project
+     */
+    NONE,
+    /**
+     * Aggregate data from all direct children. Respects collection logic of
+     * direct children collections.
+     */
+    AGGREGATE_DIRECT_CHILDREN,
+    /**
+     * Aggregate all direct children which have a specific tag
+     */
+    AGGREGATE_DIRECT_CHILDREN_WITH_TAG,
+    /**
+     * Find the child with the highest SemVer version and show only that child's data.
+     * Ignores children not using SemVer versions.
+     */
+    HIGHEST_SEMVER_CHILD

```

Comment:
I think its pretty important for versioned software.
Only alternative I could think of:
One needs logic in CI scripts to identify the latest SemVer version in Dtrack after creating a version and must auto-manage tags to always append a "highest-version" tag or similar to the latest version on its own and auto-remove it for all other versions. Then you can use the "aggregate by tag" logic to achieve the same. But this requires quite some logic in CI builds.
Or could we at least offer an API to identify the latest SemVer version of a project? That would not be called as often as Metrics update, hence doing it outside SQL would be ok. Then it would be a step easier to implement above logic. 

Or even better: one could enable an optional feature that lets the project management APIs auto-tag the latest SemVer version automatically whenever projects are added/removed/reversioned. 
Would this be an option? Then we don't lose any comfort AND the info could even be reused for other potential features, while not slowing down or making more complicated the metrics update.

Reply:
> Or could we at least offer an API to identify the latest SemVer version of a project?

When dealing with versioned software like this, wouldn't it be a fair assumption that BOMs of older versions would have been uploaded earlier than those of newer versions?

At the very least, the latest version should be somewhere among the most recently uploaded BOMs, no? So you could use this endpoint in CI to pull the last, say, `5` projects with name `foo` and then do the version comparison as needed:

```shell
curl -H 'X-Api-Key: <API_KEY>' \
    'http://localhost:8080/api/v1/project?name=foo&sortName=lastBomImport&sortOrder=desc&offset=0&limit=5'
```

To be clear, I'm not strictly against supporting SemVer in DT, I'm just trying to avoid us steering into a maintenance mess once people start asking for more versioning schemes.

Reply:
> When dealing with versioned software like this, wouldn't it be a fair assumption that BOMs of older versions would have been uploaded earlier than those of newer versions?

Not necessarily.
Lets say you have 4 major versions
1.x.x, 2.x.x, 3.x.x, 4.x.x
and you support e.g. the last 3 of those due to business contracts (>=2.x.x) additionally you have active development on the latest version (4.x.x). In this case it can easily happen your upload order is for a while (first to latest):
4.1.0
4.1.1 (bugfix)
3.4.1 (backport of bugfix)
2.3.1 (backport of bugfix)
2.3.2 (additional fix in older version that doesn't apply to newer version)
...

> At the very least, the latest version should be somewhere among the most recently uploaded BOMs, no? So you could use this endpoint in CI to pull the last, say, 5 projects with name foo and then do the version comparison as needed:
Likely, yes. I would also assume for our current use cases 5 would be enough. However, idk if that would be the case for everyone. Special cases (e.g. newly using Dtrack and importing existing SBOMs of a lot of SW versions) also not considered.

Not sure whats the best strategy here...

Reply:
> I would also assume for our current use cases 5 would be enough. However, idk if that would be the case for everyone.

Nothing would prevent users to traverse the project list to the very end in order to determine the "latest" version.

> Not sure whats the best strategy here...

My suggestion would be to investigate **if** and **how** the use case can be addressed with tools we already have available, including pulling all project versions and doing the "latest" determination on the client-side (i.e. CI scripts).

Based on that, figure out what the pains / limitations are with that approach, and how a server-side implementation could help resolve that. Are the identified limitations show-stoppers or merely annoyances?

If we are **really** sure we need a server-side implementation, we will need something that covers more than vanilla SemVer: Some organizations will version their software according to Maven, others may use .NET semantics, and so on. We also need to consider that organizations may track software they did not build (i.e. container images from vendors), and thus don't have control over versioning. The `vers` specification lists a few versioning schemes here: https://github.com/package-url/purl-spec/blob/version-range-spec/VERSION-RANGE-SPEC.rst#some-of-the-known-versioning-schemes

We don't have to cover all cases from the very beginning, but the solutions should at least be extensible such that more can be added in the future.

---

More generally, my recommendation would be to decouple the version-awareness from this PR, as to not block this from getting merged.

Reply:
Ok since I somehow want to get this feature out and have limited time, lets find a way for now.
In my opinion being able to auto-point to the latest version would be very important. for me this would be like half the projects that should rely on this.

However, I see your point that an overall and more flexible versioning solution makes sense. But, I currently can't take care of another big topic due to time constraints.
So either we take the solution suggested now as a temporary solution and state in the documentation that there are plans to add more flexibility later, allowing us to replace it in a future version, or we remove this. Possibly could replace it for now with a "LATEST_CHILD_BOM_UPLOADED", which would purely rely on the timing of the uploaded BOM (however this could produce flaky/unstable results if multiple versions are developed in parallel like a new major and a patch minor version). 
Using my current solution as temporary solution shouldn't block further replacement in my opinion, and would possibly allow us to receive feedback about requirements of others for this topic, especially if we link an improvement github Issue in the docs directly.

What do you think @nscuro? What do others think?

Reply:
Sounds like a good compromise to keep it simple. will you implement that in a separate PR first?

Reply:
Yeah this should be a separate change I think. In the meantime you can either omit the "highest version" logic in this PR, or implement it with dummy logic until the other PR lands.

Reply:
I have removed the highestsemverchild logic and replaced it with a dummy aggregating children having a future isLatest property set. I created https://github.com/DependencyTrack/dependency-track/issues/4148 to get this flag done. Once https://github.com/DependencyTrack/dependency-track/issues/4148 is implemented, the logic can be activated and only requires adding a test case, can be done as part of https://github.com/DependencyTrack/dependency-track/issues/4148, I added some notice in that issue that it shall be done when doing this.

Reply:
Enabled now the logic for AGGREGATE_LATEST_VERSION_CHILDREN and added test for it.